providers/providers.go

```diff
@@ -281,21 +282,23 @@ func (pm *ProviderManager) run() {
 					log.Error("error loading known provset: ", err)
 					continue
 				}
-				var filtered []peer.ID
 				for p, t := range provs.set {
-					if time.Now().Sub(t) > ProvideValidity {
+					if now.Sub(t) > ProvideValidity {
 						delete(provs.set, p)
-					} else {
-						filtered = append(filtered, p)
 					}
 				}
-
-				provs.providers = filtered
-				if len(filtered) == 0 {
+				// have we run out of providers?
+				if len(provs.set) == 0 {
+					provs.providers = nil
 					err := pm.deleteProvSet(k)
 					if err != nil {
 						log.Error("error deleting provider set: ", err)
 					}
+				} else if len(provs.providers) != len(provs.set) {
```

Comment:
Do you want to eliminate an allocation here when the `provs.set` is not substantially smaller than `provs.providers`? Like only reallocate if len is 2x smaller otherwise reslice to 0 and reappend (example: https://play.golang.org/p/_g0C2OQZwi).

Reply:
I considered that but your example actually has a leak, we'd need to nil out the unused entries (which we could do, it's just more complex).

Reply:
It'll look like:

```go
if len(provs.set) == 0 {
	provs.providers = nil
	err := pm.deleteProvSet(k)
	if err != nil {
		log.Error("error deleting provider set: ", err)
	}
} else if len(provs.providers) < len(provs.set) {
	if len(provs.set) < cap(provs.providers)/2 {
		provs.providers = make([]peer.ID, 0, len(provs.set))
	} else {
		for i := range provs.providers[len(provs.set):] {
			provs.providers[i] = ""
		}
		provs.providers = provs.providers[:0]
	}
	for p := range provs.set {
		provs.providers = append(provs.providers, p)
	}
}
```

As I don't think this one allocation will make much difference in the grand scheme of things, I'm inclined to leave it as it is.